File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/StatusString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class StatusString {
    private static final String DNCString = "?DNC?";

    public String code;

    public StatusString(String code) {
        this.code = code;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static StatusString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "Status");
        return new StatusString(parts.get(0));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof StatusString)) return false;
        StatusString that = (StatusString) o;
        return (DNCString.equals(code) || DNCString.equals(that.code) || Objects.equals(code, that.code));
    }

    @Override
    public int hashCode() {
        return Objects.hash(code);
    }

    @Override
    public String toString() {
        return Tokens.token(code);
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/StatusString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class StatusString
    {
        public string code;

        public StatusString(string code)
        {
            this.code = code;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static StatusString FromText(string text)
        {
            var parts = Tokens.Require(text, 1, "Status");
            return new StatusString(parts[0]);
        }

        public StatusTyped ToStatusTyped()
        {
            return new StatusTyped(
                int.Parse(this.code)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(code);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not StatusString other) return false;
            return (DNCString == this.code || DNCString == other.code || object.Equals(this.code, other.code));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.code);
            return h.ToHashCode();
        }
    }
}
